Add Paymaster, PaymasterERC20, PaymasterERC20Guarantor, PaymasterERC721Owner and PaymasterSigner (#6576)
OpenZeppelin/openzeppelin-contracts: docs/modules/ROOT/pages/paymasters.adoc

In the context of account abstraction, sponsoring user operations allows a third party to pay for transaction gas fees on behalf of users. This can improve user experience by eliminating the need for users to hold native cryptocurrency (like ETH) to pay for transactions.

To enable sponsorship, users sign their user operations including a special field called `paymasterAndData`, resulting from the concatenation of the paymaster address they're intending to use and the associated calldata that's going to be passed into xref:api:utils/cryptography.adoc#Paymaster-validatePaymasterUserOp[`validatePaymasterUserOp`]. The EntryPoint will use this field to determine whether it is willing to pay for the user operation or not.

== Signed Sponsorship

The xref:api:account.adoc#PaymasterSigner[`PaymasterSigner`] implements signature-based sponsorship via authorization signatures, allowing designated paymaster signers to authorize and sponsor specific user operations without requiring users to hold native ETH.

> TIP: Learn more about xref:accounts.adoc#selecting_a_signer[signers] to explore different approaches to user operation sponsorship via signatures.

Author: Fixed in 295f64b9: changed `selecting_a_signer` to `selecting-a-signer` to match the published Asciidoctor anchor format.

The guarantor pre-funds the maximum possible gas cost upfront, and after execution:

1. If the user repays the guarantor, the guarantor gets their funds back
2. If the user fails to repay, the guarantor absorbs the cost

[TIP]
====
A common use case is for guarantors to pay for operations of users claiming airdrops:

* The guarantor pays gas fees upfront
* The user claims their airdrop tokens
* The user repays the guarantor from the claimed tokens
* If the user fails to repay, the guarantor absorbs the cost
====

To implement guarantor functionality, your paymaster needs to extend the PaymasterERC20Guarantor class and implement the `_fetchGuarantor` function:

[source,solidity]
----
function _fetchGuarantor(
    PackedUserOperation calldata userOp
) internal view override returns (address guarantor) {
    // Implement logic to fetch and validate the guarantor from userOp
}
----

Let's create a guarantor-enabled paymaster by extending our previous example:

> ```solidity

Author: Fixed in 295f64b9: replaced the Markdown triple-backtick fence with AsciiDoc `[source,solidity]` + `----` delimiters, consistent with the other source blocks on the page.

// WARNING: Unaudited code.
// Consider performing a security review before going to production.
contract PaymasterUSDCGuaranteed is EIP712, PaymasterERC20Guarantor, Ownable {
    
    // Keep the same oracle code as before...
    
    bytes32 private constant GUARANTEED_USER_OPERATION_TYPEHASH =
        keccak256(
            "GuaranteedUserOperation(address sender,uint256 nonce,bytes initCode,bytes callData,bytes32 accountGasLimits,uint256 preVerificationGas,bytes32 gasFees,bytes paymasterData)"
        );

    constructor(
        address initialOwner
    ) EIP712("PaymasterUSDCGuaranteed", "1") Ownable(initialOwner) {}
    
    // Other functions from PaymasterUSDCChainlink...
    
    function _fetchGuarantor(
        PackedUserOperation calldata userOp
    ) internal view override returns (address guarantor) {
        bytes calldata paymasterData = userOp.paymasterData();
        
        // Check guarantor data (should be at least 22 bytes: 20 for address + 2 for sig length)
        // If no guarantor specified, return early
        if (paymasterData.length < 22 || guarantor == address(0)) {

>             return address(0);

Author: Fixed in 295f64b9: reordered the example so `paymasterData.length` is checked first, `guarantor` is parsed from `paymasterData[:20]`, and only then `guarantor == address(0)` is validated. The original snippet read `guarantor` before assigning it, so the check was always against zero.

  domain: {
    chainId: await signerClient.getChainId(),
    name: "MyPaymasterECDSASigner",
    verifyingContract: paymasterECDSASigner.address,
    version: "1",
  },
  types: {
    UserOperationRequest: [
      { name: "sender", type: "address" },
      { name: "nonce", type: "uint256" },
      { name: "initCode", type: "bytes" },
      { name: "callData", type: "bytes" },
      { name: "accountGasLimits", type: "bytes32" },
      { name: "preVerificationGas", type: "uint256" },
      { name: "gasFees", type: "bytes32" },
      { name: "paymasterVerificationGasLimit", type: "uint256" },
      { name: "paymasterPostOpGasLimit", type: "uint256" },
      { name: "validAfter", type: "uint48" },
      { name: "validUntil", type: "uint48" },
    ],
  },
  primaryType: "UserOperationRequest",
  message: {
    sender: userOp.sender,
    nonce: userOp.nonce,
    initCode: userOp.initCode,
    callData: userOp.callData,
    accountGasLimits: userOp.accountGasLimits,
    preVerificationGas: userOp.preVerificationGas,
    gasFees: userOp.gasFees,
    paymasterVerificationGasLimit,
    paymasterPostOpGasLimit,
    validAfter,
    validUntil,
  },
});
----

> The time window (`validAfter` and `validUntil`) prevents replay attacks and allows you to limit how long the signature remains valid. Once signed, the paymaster data needs to be formatted and attached to the user operation:

Reviewer: because we are signing a userOp that has a nonce, replay protection should already be covered independantly of the time window.
```suggestion
The time window (`validAfter` and `validUntil`) allows you to limit how long the signature remains valid. Once signed, the paymaster data needs to be formatted and attached to the user operation:
```

As you can see, a `_fetchOracleDetails` function is specified to fetch the token price that will be used as a reference for calculating the final ERC-20 payment. One can fetch and process price data from Chainlink oracles to determine the exchange rate between ETH and a concrete ERC-20. An example with USDC would be:

1. Fetch the current `ETH/USD` and `USDC/USD` prices from their respective oracles.

> 2. `tokenPerNative` is the number of USDC units (smallest denomination) charged per wei of gas. Combining the two feeds: `USDC_per_wei = (ETH_USD / USDC_USD) * 10^USDC_decimals / 10^18`. Multiply by `_tokenPerNativeDenominator()` to get the fixed-point value expected by xref:api:account.adoc#PaymasterERC20-_erc20Cost-uint256-uint256-uint256-[`_erc20Cost`].

Reviewer: ```suggestion
2. `tokenPerNative` is the number of USDC units (smallest denomination) charged per native token. Combining the two feeds: `tokenPerNative = (ETH_USD / USDC_USD) * 10**USDC_decimals. Note that the oracles may come with their own denominator that has to be considered.
```

> NOTE: The price of the ERC-20 must be scaled by xref:api:account.adoc#PaymasterERC20-_tokenPerNativeDenominator--[`_tokenPerNativeDenominator`].

Reviewer: ```suggestion
NOTE: On mainnet `tokenPerNative` is expressed per ETH and not per wei. On most EVM chains, this native token includes 10¹⁸ units (wei). This can be customized by overriding xref:api:account.adoc#PaymasterERC20-_tokenPerNativeDenominator--[`_tokenPerNativeDenominator`]. 
```

Here's how an implementation of `_fetchOracleDetails` would look like using this approach:

TIP: Use xref:api:account.adoc#ERC4337Utils-combineValidationData-uint256-uint256-[`ERC4337Utils.combineValidationData`] to merge two `validationData` values.

[source,solidity]
----
// WARNING: Unaudited code.
// Consider performing a security review before going to production.

using SafeCast for *;
using ERC4337Utils for *;

uint256 private constant USDC_DECIMALS = 6;

function _fetchOracleDetails(
    PackedUserOperation calldata /* userOp */
)
    internal
    view
    virtual
    returns (uint256 validationData, uint256 tokenPerNative)
{
    (uint256 ETHUSDValidationData, int256 ETHUSD) = _fetchPrice(
        ETH_USD_ORACLE
    );
    (uint256 USDCUSDValidationData, int256 USDCUSD) = _fetchPrice(
        USDC_USD_ORACLE
    );

    if (ETHUSD <= 0 || USDCUSD <= 0) {
        // No negative prices
        return (ERC4337Utils.SIG_VALIDATION_FAILED, 0);
    }

    // USDC_units / wei = (ETH/USD) / (USDC/USD) * 10^USDC_decimals / 10^18, multiplied by

>     // `_tokenPerNativeDenominator()` to give the fixed-point value expected by `_erc20Cost`.

Reviewer: ```suggestion
    // USDC_units / eth = (ETH/USD) / (USDC/USD) * 10 ** USDC_decimals
```

    int256 scale = _tokenPerNativeDenominator().toInt256();
    int256 numerator = ETHUSD *
        (10 ** USDC_USD_ORACLE.decimals()).toInt256() *
        (10 ** USDC_DECIMALS).toInt256() *
        scale;
    int256 denominator = USDCUSD * (10 ** ETH_USD_ORACLE.decimals()).toInt256() * int256(1e18);
    int256 scaledTokenPrice = numerator / denominator;

    return (
        ETHUSDValidationData.combineValidationData(USDCUSDValidationData),
        uint256(scaledTokenPrice) // Safe upcast

>     );

Reviewer: ```suggestion
    return (
        ETHUSDValidationData.combineValidationData(USDCUSDValidationData),
        uint256(ETHUSD)
            * (10 ** USDC.decimals())
            * (10 ** USDC_USD_ORACLE.decimals())
            / uint256(USDCUSD)
            / (10 ** ETH_USD_ORACLE.decimals())
    );
```

Reviewer: example:
- `ETHUSD`: 25_248_600_000_000 (2524,86 * 1e10)
- `ETH_USD_ORACLE.decimals()`: 10 (see above)
- `USDCUSD`: 100_000_000 = (1.00 * 1e8)
- `USDC_USD_ORACLE.decimals()` = 8 (see above)
- `USDC.decimals()`: 6

=> `25_248_600_000_000 * 1e8 * 1e6 / 100_000_000 / 1e10 = 2_524_860_000` which is the correct number of units of USDC per ETH

}

function _fetchPrice(
    AggregatorV3Interface oracle
) internal view virtual returns (uint256 validationData, int256 price) {
    (
        uint80 roundId,
        int256 price_,
        ,
        uint256 timestamp,
        uint80 answeredInRound
    ) = oracle.latestRoundData();
    if (

>         price_ == 0 || // No data

Reviewer: we could check `price_ > 0` and have this function return a uint256, that way `_fetchOracleDetails` would not need to check negative values, and could work with uint256 out of the box.

        );

    constructor(
        address initialOwner
    ) EIP712("PaymasterUSDCGuaranteed", "1") Ownable(initialOwner) {}
    
    // Other functions from PaymasterUSDCChainlink...
    
    function _fetchGuarantor(
        PackedUserOperation calldata userOp
    ) internal view override returns (address guarantor) {
        bytes calldata paymasterData = userOp.paymasterData();
        
        // Check guarantor data (should be at least 22 bytes: 20 for address + 2 for sig length)
        // If no guarantor specified, return early
        if (paymasterData.length < 22) {
            return address(0);
        }
        
        guarantor = address(bytes20(paymasterData[:20]));  
        uint16 guarantorSigLength = uint16(bytes2(paymasterData[20:22]));

        if (guarantor == address(0)) {
          return address(0);
        }
        
        // Ensure the signature fits in the data
        if (paymasterData.length < 22 + guarantorSigLength) {
            return address(0);
        }
        

>         bytes calldata guarantorSignature = paymasterData[22:22 + guarantorSigLength];

Reviewer: maybe we could skip the `guarantorSigLength` part and do

```suggestion
        // If no guarantor specified, return early
        if (paymasterData.length < 20) {
            return address(0);
        }
        
        guarantor = address(bytes20(paymasterData));
        bytes calldata guarantorSignature = paymasterData[20:];
```

      { name: "nonce", type: "uint256" },
      { name: "initCode", type: "bytes" },
      { name: "callData", type: "bytes" },
      { name: "accountGasLimits", type: "bytes32" },
      { name: "preVerificationGas", type: "uint256" },
      { name: "gasFees", type: "bytes32" },
      { name: "paymasterData", type: "bytes" }
    ]
  },
  primaryType: "GuaranteedUserOperation",
  message: {
    sender: userOp.sender,
    nonce: userOp.nonce,
    initCode: userOp.initCode,
    callData: userOp.callData,
    accountGasLimits: userOp.accountGasLimits,
    preVerificationGas: userOp.preVerificationGas,
    gasFees: userOp.gasFees,
    paymasterData: guarantorAddress // Just the guarantor address
  },
});
----

Then, we include the guarantor's address and its signature in the paymaster data:

[source,typescript]
----
const paymasterVerificationGasLimit = 150_000n;
const paymasterPostOpGasLimit = 300_000n;

userOp.paymasterAndData = encodePacked(
  ["address", "uint128", "uint128", "bytes"],
  [
    paymasterUSDC.address,
    paymasterVerificationGasLimit,
    paymasterPostOpGasLimit,
    encodePacked(
      ["address", "bytes2", "bytes"],
      [
        guarantorAddress, 

>         toHex(guarantorSignature.replace("0x", "").length / 2, { size: 2 }),
>         guarantorSignature
>       ]
>     )

Reviewer: See https://github.com/OpenZeppelin/openzeppelin-contracts/pull/6576/changes#r3406424328

```suggestion
    concat([
        guarantorAddress, 
        guarantorSignature
    ])
```

Note: this is based on `ethers.concat`. Alternative names could be `bytes.concat` or `concatBytes`
